pulsar-transaction/coordinator/src/main/java/org/apache/pulsar/transaction/coordinator/impl/MLTransactionMetadataStore.java

```diff
@@ -400,6 +400,10 @@ public CompletableFuture<Void> updateTxnStatus(TxnID txnID, TxnStatus newStatus,
                             appendLogCount.increment();
                             try {
                                 synchronized (txnMetaListPair.getLeft()) {
+                                    if (txnMetaListPair.getLeft().status() == newStatus) {
```

Comment:
1. This is a great fix, but this patch doesn't solve all issues.

---

2. There already done this verification at https://github.com/apache/pulsar/blob/7e0a6c48d23a3706182977c2e47eea211306ab5a/pulsar-transaction/coordinator/src/main/java/org/apache/pulsar/transaction/coordinator/impl/MLTransactionMetadataStore.java#L385-L388

---

3. And an additional issue:  duplicate transaction logs are persisted. https://github.com/apache/pulsar/blob/7e0a6c48d23a3706182977c2e47eea211306ab5a/pulsar-transaction/coordinator/src/main/java/org/apache/pulsar/transaction/coordinator/impl/MLTransactionMetadataStore.java#L398

---

The root cause: the in-flight persistent transaction log task is ignored during the status check. 
Can we make a mechanism to avoid this problem?

Reply:
for 2:  Lines 385 to 388 can't avoid race condition, so i check the state status in synchronized block.
for 3: this can be fixed easily, we just delete the duplicate transaction logs.
```
transactionLog.deletePosition(Collections.singletonList(position));
```